lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,164 @@
+// stdout. If the input string changes, then the commands are executed one after
+// the other and the concatenation of their outputs is produced line by line.
+//
+// Note that in the likely case in which the process emits several lines one
+// after the other, the downstream resources might not run for every line unless
+// the "Meta:realize" metaparam is set to true.
+type ExecFunc struct {
+	init *interfaces.Init
+
+	// Used to shutdown the node, killing the current process if needed.
+	// We only run one process at a time.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// ArgGen returns the Nth arg name for this function.
+func (obj *ExecFunc) ArgGen(index int) (string, error) {
+	seq := []string{"shell_command"}
+	if l := len(seq); index >= l {
+		return "", fmt.Errorf("index %d exceeds arg length of %d", index, l)
+	}
+	return seq[index], nil
+}
+
+// Validate makes sure we've built our struct properly. It is usually unused for
+// normal functions that users can use directly.
+func (obj *ExecFunc) Validate() error {
+	return nil
+}
+
+// Info returns some static info about itself.
+func (obj *ExecFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: false, // definitely false
+		Memo: false,
+		Sig:  types.NewType("func(shell_command str) str"),
+		Err:  obj.Validate(),
+	}
+}
+
+// Init runs some startup code for this function.
+func (obj *ExecFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	obj.ctx, obj.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	// Close the output chan to signal that no more values are coming.
+	defer close(obj.init.Output)
+
+	// Kill the current process if it's still running.
+	defer obj.cancel()
+
+	for {
+		select {
+		case input, more := <-obj.init.Input:
+			if !more {
+				return nil
+			}
+			shellCommand := input.Struct()["shell_command"].Str()
+
+			// Run the command, connecting it to ctx so we can kill
+			// it if needed, and to two Readers so we can read its
+			// stdout and stderr.
+			cmd := exec.CommandContext(obj.ctx, "sh", "-c", shellCommand)
+			stdoutReader, err := cmd.StdoutPipe()
+			stderrReader, err := cmd.StderrPipe()
+			if err != nil {
+				return err
+			}
+			err = cmd.Start()
+			if err != nil {
+				return err
+			}
+
+			// Sends the command's stdout and stderr to the
+			// Readers. Terminates when the process exits, on its
+			// own or or due to obj.cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				cmd.Wait()
+			}()
+
+			// Log the lines from stderr, to help the user debug.
+			// Terminates when the process exits, on its own or or
+			// due to obj.cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				stderrScanner := bufio.NewScanner(stderrReader)
+				for stderrScanner.Scan() {
+					obj.init.Logf(
```

Comment:
It's okay to put all of this on one line unless you prefer it like this. Docstrings and comments usually wrap at 80 chars, but it's okay to let code lines that are sensible to run a bit longer.